lib/segment/src/payload_storage/condition_checker.rs

```diff
@@ -140,18 +137,34 @@ impl ValueChecker for Match {
     }
 }
 
-impl ValueChecker for Range<FloatPayloadType> {
+impl ValueChecker for RangeInterface {
     fn check_match(&self, payload: &Value) -> bool {
-        match payload {
-            Value::Number(num) => num
-                .as_f64()
-                .map(|number| self.check_range(number))
-                .unwrap_or(false),
-            _ => false,
+        match self {
+            RangeInterface::Float(condition) => condition.check_match(payload),
+            RangeInterface::Int(condition) => condition.check_match(payload),
+            RangeInterface::DateTime(condition) => condition.check_match(payload),
         }
     }
 }
 
+impl ValueChecker for Range<FloatPayloadType> {
+    fn check_match(&self, payload: &Value) -> bool {
+        payload
+            .as_number()
+            .and_then(|x| x.as_f64().or_else(|| x.as_i64().map(|x| x as f64)))
+            .map_or(false, |x| self.check_range(x))
+    }
+}
+
+impl ValueChecker for Range<IntPayloadType> {
+    fn check_match(&self, payload: &Value) -> bool {
+        payload
+            .as_number()
+            .and_then(|x| x.as_i64().or_else(|| x.as_f64().map(|x| x as i64)))
```

Comment:
This rounds floats down, so, e.g., `{"payload": 0.3}` would be rounded to `0i64` and then matched by `{lte: Some(0i64)}` which is not what we want.
I think a similar rounding issue could also be in the case above (`x as f64`).
I propose to do comparison using `num-cmp`, like discussed in https://github.com/qdrant/qdrant/pull/3533#pullrequestreview-1864712419. I.e. explicitly handle cases `i64<i64`, `i64<f64`, `f64<i64`, and so on.

Reply:
You're totally right, I'll update accordingly

Reply:
@xzfc I have added a new check_range implementation for the cases where we need to check different types.

For the case of `Range<f64>`, though, I've restored its original implementation, as all integers were getting converted to floats via `as f64` inside the `serde_json` crate already.